[material-ui-icons] Reduce code duplication

cherniavskii opened the pull request:
<!-- Thanks so much for your PR, your contribution is appreciated! ❤️ -->
Closes #10900 

I've added [`withSvgIcon` file](https://github.com/cherniavskii/material-ui/blob/04a55d567acd86d23d4ffcb39fb9ccb125ff2f61/packages/material-ui-icons/src/components/withSvgIcon.js) (not sure about name, maybe it should be named `createSvgIcon`?).
It receives `path` and `displayName`, and returns pure react component.

oliviertassinari commented:
Yeah, it's not an HOC. I would rather call it `createSvgIcon`.
Have you thought of removing the React import too?

oliviertassinari commented:
> Have you thought of removing the React import too?

OK no. Not a good idea. We have no guarantee of the svg structure.

oliviertassinari commented:
We can't use `React.Fragment` as we still support React 15, nor we should need it.

cherniavskii commented:
>Any particular reason for replacing a stateless functional component with a class?

Yes, to make that component pure.
`recompose/pure` was used before, but it's just a HOC, which wraps component into a `PureComponent`.
So I decided to use `PureComponent` directly

cherniavskii commented:
@oliviertassinari fair enough.

But I'm wondering why `PhotoCamera` icon works with React 15?
There are two nodes (circle and path), which are not supported by React 15 :/
Will check this

mbrookes commented:
> Yes, to make that component pure.

It already is.

> recompose/pure [is] just a HOC, which wraps component into a PureComponent.

Not true. Trust me when I say this is a well flogged horse. Here, and [elsewhere](https://github.com/gorangajic/react-icons/issues/80#issuecomment-273340749) 😜 

Not only are you dragging along unneeded baggage in the class (state management, lifecycle methods etc.), PureComponent is doing an unneeded check for state change when there is no state.

cherniavskii commented:
@mbrookes wow, that's new!
I was sure `recompose/pure` uses `React.PureComponent` under the hood. Good to know it's not true :D

I'll use `recompose/pure` instead of `React.PureComponent` then :)

leMaik commented:
@cherniavskii 

> But I'm wondering why PhotoCamera icon works with React 15?

A component can have multiple children since React 0.x, so `<svg><path ... /><path ... /><circle /></svg>` is fine.

cherniavskii commented:
@leMaik right, I mixed up multiple children and returning multiple elements from component :D

cherniavskii commented:
Ok, I've:
- replaced `React.PureComponent` with `recompose/pure`  - here's [implementation](https://github.com/cherniavskii/material-ui/blob/mui-icons-add-factory/packages/material-ui-icons/src/utils/createSvgIcon.js) 
- renamed `withSvgIcon` to `createSvgIcon` and moved it to `src/utils` folder
- replaced `React.Fragment` with `<g>`, which resolves multiple svg children issue - here's [template](https://github.com/cherniavskii/material-ui/blob/mui-icons-add-factory/packages/material-ui-icons/tpl/SvgIcon.js)

oliviertassinari commented:
@cherniavskii Let me know what you think about the current state of the pull-request.

cherniavskii commented:
@oliviertassinari I think we should export `createSvgIcon` in `index.js` file

oliviertassinari commented:
@cherniavskii What would be the use case? On the docs, we have been writing our own icons: - 
- https://github.com/mui-org/material-ui/blob/8aa2ff2a8b41c1681a896734f7482ddd7ea4f7ca/packages/material-ui-docs/src/svgIcons/GitHub.js.
- etc.

cherniavskii commented:
@oliviertassinari we can use `createSvgIcon` instead, and allow users to use it for custom icons creation - they'll have custom icons with the same API as `material-ui-icons`

mbrookes commented:
This PR might be a good opportunity to publish under the new namespace. At some point we could add a deprecation warning to the old package.

oliviertassinari commented:
> we can use createSvgIcon instead, and allow users to use it for custom icons creation - they'll have custom icons with the same API as material-ui-icons

@cherniavskii Why not, but I don't think that we should hold this pull-request for it. We have been living without it for years. People can do a direct import of the private module (now).

> This PR might be a good opportunity to publish under the new namespace.

@mbrookes Yeah, we can trade one feature for a breaking change. I like that 👍.

oliviertassinari commented:
@cherniavskii Thank you!
